File: src/main/java/com/lessmarkup/engine/module/ModuleIntegrationImpl.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.lessmarkup.engine.module;

import com.lessmarkup.interfaces.data.DataObject;
import com.lessmarkup.interfaces.module.*;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

@Implements(ModuleIntegration.class)
public class ModuleIntegrationImpl implements ModuleIntegration {

    private final Map<Class<?>, EntitySearch> entitySearch = new HashMap<>();
    private final List<UserPropertyProvider> userPropertyProviders = new LinkedList<>();
    private String registerindModuleType;
    private final Map<String, ModuleActionHandler> moduleActionHandlers = new HashMap<>();
    
    @Override
    public void registerBackgroundJobHandler(BackgroundJobHandler handler) {
    }

    @Override
    public boolean doBackgroundJobs() {
        return true;
    }

    @Override
    public void registerEntitySearch(Class<? extends DataObject> type, EntitySearch entitySearch) {
        this.entitySearch.put(type, entitySearch);
    }

    @Override
    public void registerUserPropertyProvider(UserPropertyProvider provider) {
        userPropertyProviders.add(provider);
    }

    @Override
    public EntitySearch getEntitySearch(Class<? extends DataObject> collectionType) {
        return entitySearch.get(collectionType);
    }

    @Override
    public Collection<UserProperty> getUserProperties(long userId) {
        List<UserProperty> ret = new ArrayList<>();
        
        userPropertyProviders.forEach(provider -> {
            provider.getProperties(userId).forEach(ret::add);
        });
        
        return ret;
    }
    
    public String getRegisteringModuleType() {
        return registerindModuleType;
    }
    
    public void setRegisteringModuleType(String value) {
        registerindModuleType = value;
    }

    @Override
    public void registerActionHandler(String name, ModuleActionHandler handler) {
        moduleActionHandlers.put(name, handler);
    }

    @Override
    public ModuleActionHandler getActionHandler(String name) {
        return moduleActionHandlers.get(name);
    }
}

File: src/main/scala/com/lessmarkup/engine/module/ModuleIntegrationImpl.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.engine.module

import com.lessmarkup.interfaces.annotations.Implements
import com.lessmarkup.interfaces.data.DataObject
import com.lessmarkup.interfaces.module._

import scala.collection.JavaConversions._
import scala.collection.mutable

@Implements(classOf[ModuleIntegration]) class ModuleIntegrationImpl extends ModuleIntegration {
  private final val entitySearch: mutable.Map[Class[_], EntitySearch] = mutable.Map()
  private final val userPropertyProviders: mutable.ListBuffer[UserPropertyProvider] = mutable.ListBuffer()
  private var registerindModuleType: String = null
  private final val moduleActionHandlers: mutable.Map[String, ModuleActionHandler] = mutable.Map()

  def registerBackgroundJobHandler(handler: BackgroundJobHandler) {
  }

  def doBackgroundJobs(): Boolean = {
    true
  }

  def registerEntitySearch(`type`: Class[_ <: DataObject], entitySearch: EntitySearch) {
    this.entitySearch.put(`type`, entitySearch)
  }

  def registerUserPropertyProvider(provider: UserPropertyProvider) {
    userPropertyProviders.add(provider)
  }

  def getEntitySearch(collectionType: Class[_ <: DataObject]): EntitySearch = {
    entitySearch.get(collectionType).get
  }

  def getUserProperties(userId: Long): Seq[UserProperty] = {
    userPropertyProviders.flatten(provider => provider.getProperties(userId))
  }

  def getRegisteringModuleType: String = {
    registerindModuleType
  }

  def setRegisteringModuleType(value: String) {
    registerindModuleType = value
  }

  def registerActionHandler(name: String, handler: ModuleActionHandler) {
    moduleActionHandlers.put(name, handler)
  }

  def getActionHandler(name: String): ModuleActionHandler = {
    moduleActionHandlers.get(name).get
  }
}